### PR TITLE
feat(proposal): add numberOfDatasets field and sync on dataset changes with migration script

### PR DESCRIPTION
**Pull request by abdimo101**
## Description
This PR adds a `numberOfDatasets` field to the proposal schema and ensures it is incremented or decremented whenever datasets are created or deleted. It also includes a migration script to calculate and populate the `numberOfDatasets` field for existing proposals.

## Fixes

* Bug fixed (#X)

## Changes:

* changes made

## Tests included

- [ ] Included for each change/fix?
- [ ] Passing? 

## Documentation
- [ ] swagger documentation updated (required for API changes)
- [ ] official documentation updated

### CODE REVIEW

File: src/datasets/datasets.service.ts [hunk 1]
```diff
@@ -99,7 +101,16 @@ export class DatasetsService {
...
     }
-    return createdDataset.save();
+
+    const savedDataset = await createdDataset.save();
+
+    if (savedDataset.proposalIds) {
+      await this.proposalService.incrementNumberOfDatasets(
+        savedDataset.proposalIds,
+      );
+    }
```
Comment (Junjiequan) [inline src/datasets/datasets.service.ts:161]: `savedDataset.proposalIds` is true when its an empty array

File: test/DatasetV4.js [hunk 2]
```diff
@@ -353,6 +352,29 @@ describe("2500: Datasets v4 tests", () => {
...
+      };
+      await request(appUrl)
+        .post("/api/v4/datasets")
+        .send(dataset)
+        .auth(accessTokenAdminIngestor, { type: "bearer" });
+
+      const proposal = await request(appUrl)
+        .get(`/api/v3/Proposals/${encodeURIComponent(proposalId)}`)
+        .auth(accessTokenAdminIngestor, { type: "bearer" });
+      console.log("numberOfDatasets: ", JSON.stringify(proposal.body.numberOfDatasets, null, 2));
```
Comment (Junjiequan) [inline test/DatasetV4.js]: Can we add 2 more tests here, one for decrement, and the other to make sure increment do not happen when dataset has no or empty proposals

**Review by Junjiequan — CHANGES_REQUESTED** [review]

**Review by Junjiequan — APPROVED** [review]
